### PR TITLE
8312495: assert(0 <= i && i < _len) failed: illegal index after JDK-8287061 on big endian platforms

### PR DESCRIPTION
**Pull request by reinrich**
On big endian platforms `jint` values are stored in the high part of `StackValue` values. Therefore the the `StackValue` cannot be cast directly to `jint`. More details why this has to be like this are given in the JBS issue.

This is a common pattern. See also

https://github.com/openjdk/jdk/blob/8d29329138d44800ee4c0c02dacc01a06097de66/src/hotspot/share/runtime/deoptimization.cpp#L1386-L1387
https://github.com/openjdk/jdk/blob/8d29329138d44800ee4c0c02dacc01a06097de66/src/hotspot/share/runtime/deoptimization.cpp#L1513-L1514

### Testing
Manny iterations of vmTestbase/vm/mlvm/meth/stress/compiler/sequences/Test.java.

JTReg tests: tier1-4 of hotspot and jdk. All of Langtools and jaxp. Renaissance benchmarks as functional tests.

All testing was done with fastdebug and release builds on the main platforms and also on Linux/PPC64le.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312495](https://bugs.openjdk.org/browse/JDK-8312495): assert(0 &lt;= i &amp;&amp; i &lt; _len) failed: illegal index after JDK-8287061 on big endian platforms (**Bug** - P2)

### Reviewers
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**) ⚠️ Review applies to [224a65f7](https://git.openjdk.org/jdk/pull/14976/files/224a65f7cdf29d3bf5cda25c46a87bb90a033138)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14976/head:pull/14976` \
`$ git checkout pull/14976`

Update a local copy of the PR: \
`$ git checkout pull/14976` \
`$ git pull https://git.openjdk.org/jdk.git pull/14976/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14976`

View PR using the GUI difftool: \
`$ git pr show -t 14976`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14976.diff">https://git.openjdk.org/jdk/pull/14976.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14976#issuecomment-1646002467)

### CODE REVIEW

File: src/hotspot/share/code/debugInfo.cpp [hunk 1]
```diff
@@ -242,7 +242,10 @@ void ObjectValue::print_fields_on(outputStream* st) const {
 // of the object.
 ObjectValue* ObjectMergeValue::select(frame& fr, RegisterMap& reg_map) {
   StackValue* sv_selector = StackValue::create_stack_value(&fr, &reg_map, _selector);
-  jint selector = sv_selector->get_int();
+  // On big endian platforms the jint is in the high part of the StackValue
+  intptr_t val = sv_selector->get_int();
+  jint selector = (jint)*((jint*)&val);
+
```
Comment (RealCLanger) [inline src/hotspot/share/code/debugInfo.cpp:248]: not necessary to add a new line here

**Review by RealCLanger — APPROVED** [review]
Looks good. Minor formatting comment.

**Comment by dean-long** [comment]
How about introducing something like `jint Value::get_jint()` that can be used in place of this error-prone pattern?

File: src/hotspot/share/code/debugInfo.cpp [hunk 2]
```diff
@@ -242,7 +242,10 @@ void ObjectValue::print_fields_on(outputStream* st) const {
 // of the object.
 ObjectValue* ObjectMergeValue::select(frame& fr, RegisterMap& reg_map) {
   StackValue* sv_selector = StackValue::create_stack_value(&fr, &reg_map, _selector);
-  jint selector = sv_selector->get_int();
+  // On big endian platforms the jint is in the high part of the StackValue
+  intptr_t val = sv_selector->get_int();
+  jint selector = (jint)*((jint*)&val);
```
Comment (vnkozlov) [inline src/hotspot/share/code/debugInfo.cpp:247]: You don't need outer cast but you may need `()` around `(&val)`. See the example:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/stackValueCollection.cpp#L29

**Comment by vnkozlov** [comment]
> How about introducing something like `jint Value::get_jint()` that can be used in place of this error-prone pattern?

And may be rename current method `get_int()` to `get_intptr()` to match type of return value.

**Comment by reinrich** [comment]
Thanks for the feedback!

> > How about introducing something like `jint Value::get_jint()` that can be used in place of this error-prone pattern?
> 
> And may be rename current method `get_int()` to `get_intptr()` to match type of return value.

I agree. I thought it should be done in a dedicated RFE. But I can do it in this PR.

**Comment by dean-long** [comment]
> I agree. I thought it should be done in a dedicated RFE. But I can do it in this PR.

It seems reasonable to me to allow small enhancements along with bug fixes.  If it's a separate RFE then you have to either push that first, or do a temporary bug fix that doesn't use the enhancement.  What are the chances that we would want to back-port one without the other?

**Comment by reinrich** [comment]
I've done the the refactoring. There are a few places were I wasn't sure what to do:

* differences in the handling of `Location::dbl` and `lng` in `reassign_fields_by_klass` and `reassign_type_array_elements`
* Should the logic of `StackValueCollection::float_at` and similar be moved to `StackValue`?
* what to do with `byte_array_put`?
* Should `get_jint` be implemented as a simple (jint) cast of `_integer_value` on little endian platforms?

**Comment by vnkozlov** [comment]
> I've done the the refactoring. There are a few places were I wasn't sure what to do:

Good.

> * differences in the handling of `Location::dbl` and `lng` in `reassign_fields_by_klass` and `reassign_type_array_elements`

What difference you talking about?

> * Should the logic of `StackValueCollection::float_at` and similar be moved to `StackValue`?

I think this is the case for separate RFE.

> * what to do with `byte_array_put`?

Explain please

> * Should `get_jint` be implemented as a simple (jint) cast of `_integer_value` on little endian platforms?

No. The code you suggest works for all case - there is no need to complicate it.

**Comment by reinrich** [comment]
The langtools/tier1 failure on linux-x86 looks unrelated

```
#  Internal Error (g1ConcurrentMark.cpp:1671), pid=37061, tid=37071
#  fatal error: Overflow during reference processing, can not continue. Current mark stack depth: 65472, MarkStackSize: 65536, MarkStackSizeMax: 4194304. Please increase MarkStackSize and/or MarkStackSizeMax and restart.
```

**Comment by reinrich** [comment]
> 
> > * differences in the handling of `Location::dbl` and `lng` in `reassign_fields_by_klass` and `reassign_type_array_elements`
> 
> What difference you talking about?

`reassign_fields_by_klass` falls through to the `T_LONG` and `T_DOUBLE` case

https://github.com/openjdk/jdk/blob/8008e27c55030b397e2040bc3cf8408e47edf412/src/hotspot/share/runtime/deoptimization.cpp#L1518

where the intptr is fed as-is into a simple `long_field_put`

https://github.com/openjdk/jdk/blob/2bdfa836adbeba3319bee4ee61017907d6d84d58/src/hotspot/share/runtime/deoptimization.cpp#L1528

`reassign_type_array_elements` extracts the halfs from the intptr value and stores them separately

https://github.com/openjdk/jdk/blob/8008e27c55030b397e2040bc3cf8408e47edf412/src/hotspot/share/runtime/deoptimization.cpp#L1383-L1384

Instead of the 2 stores with the complex casts just one `long_at_put` could be done. Or is there a reason not to do that?

> > * Should the logic of `StackValueCollection::float_at` and similar be moved to `StackValue`?
> 
> I think this is the case for separate RFE.

Ok.

> > * what to do with `byte_array_put`?
> 
> Explain please

Well, it has very similar patterns like the locations I changed so far. I guess it is better to leave it alone for now.

> > * Should `get_jint` be implemented as a simple (jint) cast of `_integer_value` on little endian platforms?
> 
> No. The code you suggest works for all case - there is no need to complicate it.

Sure.

**Comment by vnkozlov** [comment]
> The langtools/tier1 failure on linux-x86 looks unrelated
> 
> ```
> #  Internal Error (g1ConcurrentMark.cpp:1671), pid=37061, tid=37071
> #  fatal error: Overflow during reference processing, can not continue. Current mark stack depth: 65472, MarkStackSize: 65536, MarkStackSizeMax: 4194304. Please increase MarkStackSize and/or MarkStackSizeMax and restart.
> ```

https://bugs.openjdk.org/browse/JDK-8312534

**Comment by vnkozlov** [comment]
> > > * differences in the handling of `Location::dbl` and `lng` in `reassign_fields_by_klass` and `reassign_type_array_elements`
> > 
> > 
> > What difference you talking about?

So you are asking about difference for `case T_INT: case T_FLOAT:`  and not `case T_LONG: case T_DOUBLE:`.
Let leave it as it is for now and file RFE to investigate (to factor out and use the same code).

**Comment by vnkozlov** [comment]
>>> what to do with byte_array_put?
>> Explain please
> Well, it has very similar patterns like the locations I changed so far. I guess it is better to leave it alone for now.

It is not big change - just pass `StackValue*` argument.

**Comment by reinrich** [comment]
> > > > * differences in the handling of `Location::dbl` and `lng` in `reassign_fields_by_klass` and `reassign_type_array_elements`
> > > 
> > > 
> > > What difference you talking about?
> 
> So you are asking about difference for `case T_INT: case T_FLOAT:` and not `case T_LONG: case T_DOUBLE:`. Let leave it as it is for now and file RFE to investigate (to factor out and use the same code).

Ok. I've created https://bugs.openjdk.org/browse/JDK-8312753

**Comment by reinrich** [comment]
Tests succeeded with the latest version.
@vnkozlov @dean-long are you ok with the refactoring?
I'll be out of office next week. It would be good if big endian platforms could be fixed before.
If in doubt I would do the minimal fix and postpone the refactoring.

**Review by dean-long — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]
My testing passed clean.

**Pull request closed without merging**